neuralmonkey/dataset.py

```diff
@@ -0,0 +1,559 @@
+                s_name: lambda n=s_name: iter(data[n])  # type: ignore
+                for s_name in self.iterators}
+
+    def __len__(self) -> int:
+        """Get the length of the dataset.
+
+        Returns:
+            The length of the dataset.
+
+        Raises:
+            `NotImplementedError` when the dataset is lazy.
+        """
+        if self.lazy:
+            raise NotImplementedError("Querying the len of a lazy dataset.")
+        assert self.length is not None
+        return self.length
+
+    @property
+    def series(self) -> List[str]:
+        return list(sorted(self.iterators.keys()))
+
+    def has_series(self, name: str) -> bool:
+        """Check if the dataset contains a series of a given name.
+
+        Arguments:
+            name: Series name
+
+        Returns:
+            True if the dataset contains the series, False otherwise.
+        """
+        return name in self.iterators
+
+    def get_series(self, name: str) -> Iterator:
+        """Get the data series with a given name.
+
+        Arguments:
+            name: The name of the series to fetch.
+
+        Returns:
+            A freshly initialized iterator over the data series.
+
+        Raises:
+            KeyError if the series does not exists.
+        """
+        return self.iterators[name]()
+
+    def maybe_get_series(self, name: str) -> Optional[Iterator]:
+        """Get the data series with a given name, if it exists.
+
+        Arguments:
+            name: The name of the series to fetch.
+
+        Returns:
+            The data series or None if it does not exist.
+        """
+        if name in self.iterators:
+            return self.get_series(name)
+        return None
+
+    def batches(self, batch_size: int) -> Iterator["Dataset"]:
+        """Split the dataset into batches.
+
+        Arguments:
+            batch_size: The size of a batch.
+
+        Returns:
+            Generator yielding the batches.
+        """
+        if self.buffer_size is not None and self.buffer_size < batch_size:
+            warn("Buffer size ({}) is lower than batch size ({})."
+                 .format(self.buffer_size, batch_size))
+
+        # Initialize iterators
+        iterators = {s: it() for s, it in self.iterators.items()}
+
+        # Create iterator over instances
+        zipped_iterator = (
+            dict(zip(iterators, row)) for row in zip(*iterators.values()))
+
+        # Fill the buffer with initial values, shuffle optionally
+        if self.buffer_size is not None:
+            # pylint: disable=stop-iteration-return
+            # This is pylint issue https://github.com/PyCQA/pylint/issues/2158
+            buf = deque(next(zipped_iterator) for _ in range(self.buffer_size))
+            # pylint: enable=stop-iteration-return
+        else:
+            assert not self.lazy
+            buf = deque(zipped_iterator)
+        if self.shuffled:
+            random.shuffle(buf)  # type: ignore
```

Comment:
Není efektivnější to lineárně převést na list, pak to lineárně zashuflovat a udělat z toho zase deque? Nebo je se to děje vevnitři v tom shufflu?

Reply:
Neděje. Viz co jsem napsal k Dušanovo review. Nejradši bych tam dal minimum size, pod kterou když to vleze, tak se to zashuffluje - klidně převedením na list a zpět. Ale nebude se to shufflovat po každý batchi

Reply:
snad hotovo. uvidime jestli projdou

Reply:
tests